// element-plus/lib/components/tag/src/tag2.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var index = require('../../icon/index.js');
var iconsVue = require('@element-plus/icons-vue');
require('../../../hooks/index.js');
var tag = require('./tag.js');
var pluginVue_exportHelper = require('../../../_virtual/plugin-vue_export-helper.js');
var index$1 = require('../../../hooks/use-common-props/index.js');

const _sfc_main = vue.defineComponent({
  name: "ElTag",
  components: { ElIcon: index.ElIcon, Close: iconsVue.Close },
  props: tag.tagProps,
  emits: tag.tagEmits,
  setup(props, { emit }) {
    const tagSize = index$1.useSize();
    const classes = vue.computed(() => {
      const { type, hit, effect, closable } = props;
      return [
        "el-tag",
        closable && "is-closable",
        type ? `el-tag--${type}` : "",
        tagSize.value ? `el-tag--${tagSize.value}` : "",
        effect ? `el-tag--${effect}` : "",
        hit && "is-hit"
      ];
    });
    const handleClose = (event) => {
      event.stopPropagation();
      emit("close", event);
    };
    const handleClick = (event) => {
      emit("click", event);
    };
    return {
      classes,
      handleClose,
      handleClick
    };
  }
});
const _hoisted_1 = { class: "el-tag__content" };
const _hoisted_2 = { class: "el-tag__content" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_close = vue.resolveComponent("close");
  const _component_el_icon = vue.resolveComponent("el-icon");
  return !_ctx.disableTransitions ? (vue.openBlock(), vue.createElementBlock("span", {
    key: 0,
    class: vue.normalizeClass(_ctx.classes),
    style: vue.normalizeStyle({ backgroundColor: _ctx.color }),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    vue.createElementVNode("span", _hoisted_1, [
      vue.renderSlot(_ctx.$slots, "default")
    ]),
    _ctx.closable ? (vue.openBlock(), vue.createBlock(_component_el_icon, {
      key: 0,
      class: "el-tag__close",
      onClick: _ctx.handleClose
    }, {
      default: vue.withCtx(() => [
        vue.createVNode(_component_close)
      ]),
      _: 1
    }, 8, ["onClick"])) : vue.createCommentVNode("v-if", true)
  ], 6)) : (vue.openBlock(), vue.createBlock(vue.Transition, {
    key: 1,
    name: "el-zoom-in-center"
  }, {
    default: vue.withCtx(() => [
      vue.createElementVNode("span", {
        class: vue.normalizeClass(_ctx.classes),
        style: vue.normalizeStyle({ backgroundColor: _ctx.color }),
        onClick: _cache[1] || (_cache[1] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
      }, [
        vue.createElementVNode("span", _hoisted_2, [
          vue.renderSlot(_ctx.$slots, "default")
        ]),
        _ctx.closable ? (vue.openBlock(), vue.createBlock(_component_el_icon, {
          key: 0,
          class: "el-tag__close",
          onClick: _ctx.handleClose
        }, {
          default: vue.withCtx(() => [
            vue.createVNode(_component_close)
          ]),
          _: 1
        }, 8, ["onClick"])) : vue.createCommentVNode("v-if", true)
      ], 6)
    ]),
    _: 3
  }));
}
var Tag = /* @__PURE__ */ pluginVue_exportHelper["default"](_sfc_main, [["render", _sfc_render]]);

exports["default"] = Tag;
